File: internal/gateway/classify.go
package gateway

// A rule with an intent (provider.Rule.Intent) is for a kind of message:
// as a user's turn begins, the group's classifier model is asked which of
// the intents the user's message is — once, before the turn goes anywhere,
// and never again within it. The call goes through magpie itself, so the
// classifier is any model magpie has, with its keys, failover and usage;
// it shows in the usage as magpie's own call. When it can't say
// (it fails, is slow, or answers something else) no intent matches, and
// the trace tells why.

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/yetone/magpie/internal/provider"
)

// RouterAgent is the User-Agent magpie's own classifier calls carry.
const RouterAgent = "magpie-router/1"

var (
	classifyTimeout = 8 * time.Second
	classifyKeep    = 10 * time.Minute // an answer, for the same message and intents
	classifyRest    = 30 * time.Second // after a failure, before the classifier is asked again
)

// classifier asks a model which of intents text is: the intent, or "" for
// none of them.
type classifier func(model string, intents []string, text string) (string, error)

var classified = struct {
	sync.Mutex
	m      map[string]classifiedAs
	failed map[string]classifyFailure // classifier model → its last failure
}{m: map[string]classifiedAs{}, failed: map[string]classifyFailure{}}

type classifiedAs struct {
	intent string
	at     time.Time
}

type classifyFailure struct {
	err string
	at  time.Time
}

// answerError is a classifier that answered, but not with one of the
// numbers: it is up, so it isn't left to rest for it.
type answerError struct{ error }

// classify is ask's answer for the message, from what was asked before
// when it can be: the same message and intents within classifyKeep. A
// classifier that failed is left to rest for classifyRest rather than
// making every turn wait out its timeout.
func classify(ask classifier, model string, intents []string, text string) (intent string, cached bool, err error) {
	h := sha256.New()
	h.Write([]byte(model + "\x00" + strings.ToLower(strings.Join(intents, "\x00")) + "\x00" + text))
	key := hex.EncodeToString(h.Sum(nil))
	now := time.Now()
	classified.Lock()
	if c, ok := classified.m[key]; ok && now.Sub(c.at) < classifyKeep {
		classified.Unlock()
		return c.intent, true, nil
	}
	if f, ok := classified.failed[model]; ok && now.Sub(f.at) < classifyRest {
		classified.Unlock()
		return "", false, fmt.Errorf("%s failed %s ago (%s); not asked again for now", model, now.Sub(f.at).Round(time.Second), f.err)
	}
	classified.Unlock()
	intent, err = ask(model, intents, text)
	classified.Lock()
	defer classified.Unlock()
	if err != nil {
		if !errors.As(err, new(answerError)) {
			classified.failed[model] = classifyFailure{err: err.Error(), at: time.Now()}
		}
		return "", false, err
	}
	delete(classified.failed, model)
	classified.m[key] = classifiedAs{intent: intent, at: time.Now()}
	if len(classified.m) > 4096 {
		for k, c := range classified.m {
			if time.Since(c.at) > classifyKeep {
				delete(classified.m, k)
			}
		}
	}
	return intent, false, nil
}

var reminders = regexp.MustCompile(`(?s)<system-reminder>.*?</system-reminder>`)

// userText is what the user said to begin the turn: the last user
// message's text, without what agents add to it for the model
// (<system-reminder>…), its middle left out when long.
func userText(req *Request) string {
	var parts []string
	for i := len(req.Messages) - 1; i >= 0; i-- {
		m := req.Messages[i]
		if m.Role != "user" {
			continue
		}
		for _, p := range m.Parts {
			if p.Kind == Text {
				parts = append(parts, p.Text)
			}
		}
		break
	}
	text := strings.TrimSpace(reminders.ReplaceAllString(strings.Join(parts, "\n"), ""))
	const head, tail = 3000, 1000
	if r := []rune(text); len(r) > head+tail {
		text = string(r[:head]) + "\n…\n" + string(r[len(r)-tail:])
	}
	return text
}

const classifyPrompt = "You route a user's message to a coding assistant by what it asks for. " +
	"Given numbered kinds of request and the user's message, answer with the number of the kind the message is, " +
	"or 0 if it is none of them. Answer with the number only."

// classifyBody is the Chat request asking model which of intents text is,
// at effort when it isn't "". A model that reasons does so before it
// answers, and how long varies: 2048 leaves room for that and the number.
func classifyBody(model, effort string, intents []string, text string) []byte {
	var b strings.Builder
	b.WriteString("Kinds:\n")
	for i, in := range intents {
		fmt.Fprintf(&b, "%d. %s\n", i+1, in)
	}
	b.WriteString("\nThe user's message:\n<message>\n")
	b.WriteString(text)
	b.WriteString("\n</message>\n\nThe number of its kind (0 for none):")
	req := map[string]any{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": classifyPrompt},
			{"role": "user", "content": b.String()},
		},
		"stream":      false,
		"temperature": 0,
		"max_tokens":  2048,
	}
	if effort != "" {
		req["reasoning_effort"] = effort
	}
	body, _ := json.Marshal(req)
	return body
}

var firstNumber = regexp.MustCompile(`\d+`)

// readIntent is the intent a classifier's answer names: its first number,
// 0 for none.
func readIntent(answer string, intents []string) (string, error) {
	n, err := strconv.Atoi(firstNumber.FindString(answer))
	if err != nil || n < 0 || n > len(intents) {
		a := []rune(strings.TrimSpace(answer))
		if len(a) > 80 {
			a = append(a[:80], '…')
		}
		return "", answerError{fmt.Errorf("it answered %q, not a number from 0 to %d", string(a), len(intents))}
	}
	if n == 0 {
		return "", nil
	}
	return intents[n-1], nil
}

// classifyEffort is the least reasoning model takes: none when it can go
// without, else its lowest level; "" when its levels aren't known, which
// leaves it to the vendor.
func classifyEffort(model string) string {
	p, m, ok := provider.Resolve(model)
	if !ok {
		return ""
	}
	levels := p.Efforts(m)
	if len(levels) == 0 {
		return ""
	}
	return fitEffort("none", levels)
}

// askClassifier asks model through the gateway itself, as a client would.
func (s *Server) askClassifier(model string, intents []string, text string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), classifyTimeout)
	defer cancel()
	r, err := http.NewRequestWithContext(ctx, http.MethodPost, "http://magpie/v1/chat/completions", nil)
	if err != nil {
		return "", err
	}
	r.Header.Set("Content-Type", "application/json")
	r.Header.Set("User-Agent", RouterAgent)
	w := httptest.NewRecorder()
	s.serve(w, r, provider.Chat, classifyBody(model, classifyEffort(model), intents, text))
	if ctx.Err() != nil {
		return "", fmt.Errorf("%s gave no answer in %s", model, classifyTimeout)
	}
	var out struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
		Error *struct {
			Message string `json:"message"`
		} `json:"error"`
	}
	if err := json.Unmarshal(w.Body.Bytes(), &out); err != nil {
		return "", fmt.Errorf("%s: %d, not an answer", model, w.Code)
	}
	if w.Code >= 300 || out.Error != nil {
		msg := http.StatusText(w.Code)
		if out.Error != nil && out.Error.Message != "" {
			msg = out.Error.Message
		}
		return "", fmt.Errorf("%s: %s", model, msg)
	}
	if len(out.Choices) == 0 {
		return "", answerError{fmt.Errorf("%s gave no answer", model)}
	}
	return readIntent(out.Choices[0].Message.Content, intents)
}

File: src/grouprule/classify.rs
// A rule with an intent is for a kind of message: as a user's turn begins,
// the group's classifier model is asked which of the intents the user's
// message is — once, before the turn goes anywhere, and never again within
// it. The call goes through magpie itself, so the classifier is any model
// magpie has, with its keys, failover and usage; it shows in the usage as
// magpie's own call. When it can't say (it fails, is slow, or answers
// something else) no intent matches, and the trace tells why.

use std::{
    fmt::Write as _,
    sync::{LazyLock, Mutex},
    time::{Duration, Instant},
};

use anyhow::{Context, Result, bail};
use futures_util::future::BoxFuture;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};

use super::Classifier;

// RouterAgent is the User-Agent magpie's own classifier calls carry.
pub const ROUTER_AGENT: &str = "magpie-router/1";

const CLASSIFY_TIMEOUT: Duration = Duration::from_secs(8);
const CLASSIFY_KEEP: Duration = Duration::from_secs(10 * 60); // an answer, for the same message and intents
const CLASSIFY_REST: Duration = Duration::from_secs(30); // after a failure, before the classifier is asked again
const MAX_CLASSIFIED: usize = 4096;

const CLASSIFY_PROMPT: &str = "You route a user's message to a coding assistant by what it asks for. \
Given numbered kinds of request and the user's message, answer with the number of the kind the message is, \
or 0 if it is none of them. Answer with the number only.";

static CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
    reqwest::Client::builder()
        .no_proxy()
        .build()
        .expect("build classifier HTTP client")
});

#[derive(Clone)]
struct ClassifiedAs {
    intent: String,
    at: Instant,
}

struct ClassifyFailure {
    error: String,
    at: Instant,
}

static ANSWERED: LazyLock<
    Mutex<(
        std::collections::HashMap<String, ClassifiedAs>,
        std::collections::HashMap<String, ClassifyFailure>,
    )>,
> = LazyLock::new(|| {
    Mutex::new((
        std::collections::HashMap::new(),
        std::collections::HashMap::new(),
    ))
});

// answerError is a classifier that answered, but not with one of the
// numbers: it is up, so it isn't left to rest for it.
#[derive(Debug)]
pub struct AnswerError(String);

impl std::fmt::Display for AnswerError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for AnswerError {}

fn answer_error(message: impl Into<String>) -> anyhow::Error {
    anyhow::Error::new(AnswerError(message.into()))
}

fn hex(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        let _ = write!(out, "{byte:02x}");
    }
    out
}

// classify is ask's answer for the message, from what was asked before
// when it can be: the same message and intents within CLASSIFY_KEEP. A
// classifier that failed is left to rest for CLASSIFY_REST rather than
// making every turn wait out its timeout.
pub async fn classify(
    ask: &dyn Classifier,
    model: &str,
    intents: &[String],
    text: &str,
) -> Result<(String, bool)> {
    let mut hasher = Sha256::new();
    hasher.update(model.as_bytes());
    hasher.update([0]);
    hasher.update(intents.join("\x00").to_ascii_lowercase().as_bytes());
    hasher.update([0]);
    hasher.update(text.as_bytes());
    let key = hex(&hasher.finalize());
    let now = Instant::now();
    {
        let state = ANSWERED
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());
        if let Some(saved) = state
            .0
            .get(&key)
            .filter(|saved| now.saturating_duration_since(saved.at) < CLASSIFY_KEEP)
        {
            return Ok((saved.intent.clone(), true));
        }
        if let Some(failed) = state
            .1
            .get(model)
            .filter(|failed| now.saturating_duration_since(failed.at) < CLASSIFY_REST)
        {
            let ago = now.saturating_duration_since(failed.at).as_secs();
            bail!(
                "{model} failed {ago}s ago ({}); not asked again for now",
                failed.error
            );
        }
    }
    let intent = ask.ask(model, intents, text).await;
    let mut state = ANSWERED
        .lock()
        .unwrap_or_else(|poisoned| poisoned.into_inner());
    match intent {
        Ok(intent) => {
            state.1.remove(model);
            state.0.insert(
                key,
                ClassifiedAs {
                    intent: intent.clone(),
                    at: Instant::now(),
                },
            );
            if state.0.len() > MAX_CLASSIFIED {
                state.0.retain(|_, saved| {
                    Instant::now().saturating_duration_since(saved.at) <= CLASSIFY_KEEP
                });
            }
            Ok((intent, false))
        }
        Err(error) => {
            if error.downcast_ref::<AnswerError>().is_none() {
                state.1.insert(
                    model.to_owned(),
                    ClassifyFailure {
                        error: error.to_string(),
                        at: Instant::now(),
                    },
                );
            }
            Err(error)
        }
    }
}

// classifyBody is the Chat request asking model which of intents text is,
// at effort when it isn't "". A model that reasons does so before it
// answers, and how long varies: 2048 leaves room for that and the number.
pub fn classify_body(model: &str, effort: &str, intents: &[String], text: &str) -> Value {
    let mut kinds = String::from("Kinds:\n");
    for (index, intent) in intents.iter().enumerate() {
        let _ = writeln!(kinds, "{}. {intent}", index + 1);
    }
    let message = format!(
        "{kinds}\nThe user's message:\n<message>\n{text}\n</message>\n\nThe number of its kind (0 for none):"
    );
    let mut body = json!({
        "model": model,
        "messages": [
            {"role": "system", "content": CLASSIFY_PROMPT},
            {"role": "user", "content": message}
        ],
        "stream": false,
        "temperature": 0,
        "max_tokens": 2048
    });
    if !effort.is_empty() {
        body["reasoning_effort"] = json!(effort);
    }
    body
}

// readIntent is the intent a classifier's answer names: its first number,
// 0 for none.
pub fn read_intent(answer: &str, intents: &[String]) -> Result<String> {
    let digits = answer
        .trim_start_matches(|character: char| !character.is_ascii_digit())
        .chars()
        .take_while(|character| character.is_ascii_digit())
        .collect::<String>();
    let number = digits.parse::<i64>().ok();
    match number {
        Some(n) if n >= 0 && n as usize <= intents.len() => {
            if n == 0 {
                Ok(String::new())
            } else {
                Ok(intents[n as usize - 1].clone())
            }
        }
        _ => {
            let mut quoted = answer.trim().chars().take(80).collect::<String>();
            if answer.trim().chars().count() > 80 {
                quoted.push('…');
            }
            Err(answer_error(format!(
                "it answered {quoted:?}, not a number from 0 to {}",
                intents.len()
            )))
        }
    }
}

// fitEffort is the level of the model's own nearest the one asked for — a
// tie goes up — or the one asked for when the model's aren't known.
const EFFORT_RANK: [&str; 8] = [
    "none", "minimal", "low", "medium", "high", "xhigh", "max", "ultra",
];

pub fn fit_effort(want: &str, levels: &[String]) -> String {
    if levels.is_empty() || levels.iter().any(|level| level == want) {
        return want.to_owned();
    }
    let Some(at) = EFFORT_RANK.iter().position(|level| *level == want) else {
        return want.to_owned();
    };
    let (mut best, mut dist) = (want.to_owned(), EFFORT_RANK.len());
    for level in levels {
        let Some(index) = EFFORT_RANK.iter().position(|name| name == level) else {
            continue;
        };
        if level == "none" {
            continue;
        }
        let distance = index.abs_diff(at);
        if distance < dist || distance == dist && index > at {
            best = level.clone();
            dist = distance;
        }
    }
    best
}

// classifyEffort is the least reasoning model takes: none when it can go
// without, else its lowest level; "" when its levels aren't known, which
// leaves it to the vendor.
pub fn classify_effort(model: &str) -> String {
    let Some((provider_id, model_id)) = model.split_once('/') else {
        return String::new();
    };
    let Some(catalog_id) = crate::provider::catalog_id_for_usage(provider_id) else {
        return String::new();
    };
    let Some(model) = crate::catalog::available_models(provider_id, &catalog_id)
        .into_iter()
        .find(|entry| entry.id == model_id)
    else {
        return String::new();
    };
    if model.efforts.is_empty() {
        return String::new();
    }
    fit_effort("none", &model.efforts)
}

// GatewayClassifier asks the model through the gateway itself, as a client
// would, so the call keeps magpie's routing, keys and usage.
pub struct GatewayClassifier;

impl Classifier for GatewayClassifier {
    fn ask<'a>(
        &'a self,
        model: &'a str,
        intents: &'a [String],
        text: &'a str,
    ) -> BoxFuture<'a, Result<String>> {
        Box::pin(async move {
            let body = classify_body(model, &classify_effort(model), intents, text);
            let response = CLIENT
                .post(format!("{}/chat/completions", crate::gateway::v1_url()))
                .header(reqwest::header::USER_AGENT, ROUTER_AGENT)
                .timeout(CLASSIFY_TIMEOUT)
                .json(&body)
                .send()
                .await;
            let response = match response {
                Ok(response) => response,
                Err(error) if error.is_timeout() => bail!("{model} gave no answer in 8s"),
                Err(error) => return Err(error).context(format!("{model} did not answer")),
            };
            let status = response.status();
            let bytes = response
                .bytes()
                .await
                .with_context(|| format!("{model}: {status}, not an answer"))?;
            let body: Value = serde_json::from_slice(&bytes)
                .with_context(|| format!("{model}: {}, not an answer", status.as_u16()))?;
            if let Some(message) = body
                .pointer("/error/message")
                .and_then(Value::as_str)
                .filter(|message| !message.is_empty())
            {
                bail!("{model}: {message}");
            }
            if !status.is_success() {
                bail!("{model}: {}", status.canonical_reason().unwrap_or("failed"));
            }
            let Some(content) = body
                .pointer("/choices/0/message/content")
                .and_then(Value::as_str)
            else {
                return Err(answer_error(format!("{model} gave no answer")));
            };
            read_intent(content, intents)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn answers_with_words_around_the_number_are_read() {
        let intents = vec!["a quick question".to_owned()];
        assert_eq!(
            read_intent("Kind 1.", &intents).unwrap(),
            "a quick question"
        );
    }

    #[test]
    fn odd_answers_do_not_name_an_intent() {
        let intents = vec!["x".to_owned(), "y".to_owned()];
        let error = read_intent("maybe", &intents).unwrap_err();
        assert!(error.to_string().contains("not a number"));
        // the classifier is up for it: it isn't left to rest
        assert!(error.downcast_ref::<AnswerError>().is_some());
    }

    #[test]
    fn the_same_message_and_intents_are_kept() {
        let key = |model: &str, intents: &[String], text: &str| {
            let mut hasher = Sha256::new();
            hasher.update(model.as_bytes());
            hasher.update([0]);
            hasher.update(intents.join("\x00").to_ascii_lowercase().as_bytes());
            hasher.update([0]);
            hasher.update(text.as_bytes());
            hex(&hasher.finalize())
        };
        let intents = vec!["refactoring".to_owned()];
        let first = key("c/cls", &intents, "rename this package");
        let again = key("c/cls", &intents, "rename this package");
        assert_eq!(first, again);
        let other = key("c/cls", &intents, "rename that package");
        assert_ne!(first, other);
        // the intents' case doesn't change what was asked
        assert_eq!(
            key("c/cls", &["Refactoring".to_owned()], "rename this package"),
            first
        );
    }
}
